### ISSUE TITLE
Feature request - C compiler

### ISSUE THREAD

**Issue by jeremy-rutman** [GaloisInc/cryptol-compiler#42]
We'd like to be able to use cryptol to generate C that can be directly used on microcontrollers on which Rust is not yet implemented - is there a cryptol->C compiler in the works?

**Comment by kiniry** [comment]
Such a compiler already exists.  It is largely not part of our public
release of Cryptol.  We have used to it generate systems and components for
embedded systems development.  An example of its use is found in the public
[HARDENS](https://github.com/GaloisInc/HARDENS) project, which was
developed for the Nuclear Regulatory Commission.

**Comment by weaversa** [comment]
There is also a Cryptol -> C generation capability in [SAW](https://github.com/GaloisInc/saw-script). It is very easy to use, generates a test bench along w/ the C, and works well for individual functions.

```
$ saw
sawscript> enable_experimental
sawscript> :h codegen
Description
-----------

    codegen : String -> [String] -> String -> Term -> TopLevel ()

Generate straight-line C code for the given term using SBV.

First argument is directory path ("" for stdout) for generating files.
Second argument is the list of function names to leave uninterpreted.
Third argument is C function name.
Fourth argument is the term to generate code for. It must be a
first-order function whose arguments and result are all of type
Bit, [8], [16], [32], or [64].

sawscript> codegen "directory" [] "functionNameInC" {{ \(x : [32]) -> x + 1 }}
Creating directory "directory"..
Generating: "directory/Makefile"..
Generating: "directory/functionNameInC.h"..
Generating: "directory/functionNameInC_driver.c"..
Generating: "directory/functionNameInC.c"..
Done.
sawscript> :q
$ make -C directory
cc -Wall -O3 -DNDEBUG -fomit-frame-pointer -c functionNameInC.c -o functionNameInC.o
cc -Wall -O3 -DNDEBUG -fomit-frame-pointer -c functionNameInC_driver.c -o functionNameInC_driver.o
cc -Wall -O3 -DNDEBUG -fomit-frame-pointer functionNameInC.o functionNameInC_driver.o -o functionNameInC_driver
$ directory/functionNameInC_driver
functionNameInC(0x772a5561UL) = 0x772a5562UL
```

**Comment by kiniry** [comment]
And direct LLVM generation as well, in another private project/branch.  And
JVM generation...

**Comment by jeremy-rutman** [comment]
Thanks all, very helpful. I must have missed references to C generation in the docs.

**Comment by jeremy-rutman** [comment]
> Such a compiler already exists. It is largely not part of our public release of Cryptol. We have used to it generate systems and components for embedded systems development. An example of its use is found in the public [HARDENS](https://github.com/GaloisInc/HARDENS) project, which was developed for the Nuclear Regulatory Commission.

I've gone through a lot of the HARDENS documentation and can't seem to find a cryptol->C section, if possible can you let me know what the compiler or codegenerator name is, or any other pointer to the right place? Thanks - JR

**Comment by jeremy-rutman** [comment]
> There is also a Cryptol -> C generation capability in [SAW](https://github.com/GaloisInc/saw-script). It is very easy to use, generates a test bench along w/ the C, and works well for individual functions.
> 
> ```
> $ saw
> sawscript> enable_experimental
> sawscript> :h codegen
> Description
> -----------
> 
>     codegen : String -> [String] -> String -> Term -> TopLevel ()
> 
> Generate straight-line C code for the given term using SBV.
> 
> First argument is directory path ("" for stdout) for generating files.
> Second argument is the list of function names to leave uninterpreted.
> Third argument is C function name.
> Fourth argument is the term to generate code for. It must be a
> first-order function whose arguments and result are all of type
> Bit, [8], [16], [32], or [64].
> 
> sawscript> codegen "directory" [] "functionNameInC" {{ \(x : [32]) -> x + 1 }}
> Creating directory "directory"..
> Generating: "directory/Makefile"..
> Generating: "directory/functionNameInC.h"..
> Generating: "directory/functionNameInC_driver.c"..
> Generating: "directory/functionNameInC.c"..
> Done.
> sawscript> :q
> $ make -C directory
> cc -Wall -O3 -DNDEBUG -fomit-frame-pointer -c functionNameInC.c -o functionNameInC.o
> cc -Wall -O3 -DNDEBUG -fomit-frame-pointer -c functionNameInC_driver.c -o functionNameInC_driver.o
> cc -Wall -O3 -DNDEBUG -fomit-frame-pointer functionNameInC.o functionNameInC_driver.o -o functionNameInC_driver
> $ directory/functionNameInC_driver
> functionNameInC(0x772a5561UL) = 0x772a5562UL
> ```

This is great - I didn't see the codegen on first look at SAW. Is there a way to import a  .cry file into SAW such that I could for instance generate the C equivalent of the DES.cry  script?

**Comment by weaversa** [comment]
> This is great - I didn't see the codegen on first look at SAW. Is there a way to import a .cry file into SAW such that I could for instance generate the C equivalent of the DES.cry script?

Yep! It's just `import "path\to\DES.cry";`

**Comment by jeremy-rutman** [comment]
I tried importing DES and doing codegen from DES.encrypt but hit a 'not yet supported':

```
sawscript> import "DES.cry"
sawscript> codegen "des_c" [] "des_c" {{DES.encrypt}}

SBV->C: Not yet supported: join with ((SWord 1,s6),(SWord 1,s9))
CallStack (from HasCallStack):
  error, called at ./Data/SBV/Compilers/C.hs:108:11 in sbv-10.2-6c0cf9c528cdb45a38ec91ea9af248f20680acc36f40bf77ed55529a9baf928e:Data.SBV.Compilers.C

```
Any help getting encrypt/decrypt C code from any cryptol example would be greatly appreciated.

**Comment by jeremy-rutman** [comment]
In an attempt to get to a working poc I tried making an increment.cry file:
```
module Increment where
increment : [16] -> [16]
increment msg = msg+1:[16]
```
but sawscrupt doesn't seem to like this :
```
sawscript> import "increment.cry"
sawscript> codegen "increment" [] "increment" {{increment}}

Stack trace:
"codegen" (<stdin>:1:1-1:8)
asFirstOrderType: unsupported argument type: isort 0

```

**Comment by andrew-bivin** [comment]
@jeremy-rutman You'll hear from other folks more officially, but getting you the instructions on the proof of concept for Cryptol -> C for AES via the `libvcrypt` repo:

1. Clone https://github.com/GaloisInc/libvcrypt
2. Build saw if necessary
3. `cd Ciphers/Symmetric/AES`
4. `saw codegen.saw`

**Comment by jeremy-rutman** [comment]
Yes I actually already did this, having been invited to the libvcrypt repo and seeing the latest commits,  I tried and ‘saw’ the codegen works - thanks for that fast work, I think it’s all we need for proof of concept at this pointSent from my iPhoneOn Apr 25, 2024, at 12:35 AM, Andrew Bivin ***@***.***> wrote:﻿
@jeremy-rutman You'll hear from other folks more officially, but getting you the instructions on the proof of concept for Cryptol -> C for AES via the libvcrypt repo:

Clone https://github.com/GaloisInc/libvcrypt
Build saw if necessary
cd Ciphers/Symmetric/AES
saw codegen.saw
